frontend/src/liboperator/operator/operator.go

```diff
@@ -480,6 +485,100 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
 	proxy.ServeHTTP(w, r)
 }
 
+// WebSocket endpoint that proxies ADB
+func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+
+	if dev == nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return
+	}
+
+	devInfo := dev.privateData.(map[string]interface{})
+
+	// Find adb port for the device.
+	//   1. If registered device info has adb_port, use it.
+	//   2. If not, get index of the device and 6520 + index would be the adb port.
+	adbPort := 0
+	if adb_port, ok := devInfo["adb_port"]; ok {
+		adbPort = int(adb_port.(float64))
+	} else {
+		for i, id := range pool.DeviceIds() {
+			if id == devId {
+				adbPort = 6520 + i
+				break
+			}
+		}
+	}
+	if adbPort == 0 {
+		http.Error(w, "Cannot find adb port for the device", http.StatusNotFound)
+		return
+	}
+
+	// Prepare WebSocket and TCP socket for ADB
+	upgrader := websocket.Upgrader{}
+	wsConn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Print("Error while upgrading to WebSocket: ", err)
+		return
+	}
+	defer wsConn.Close()
+
+	tcpConn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%v", adbPort))
+	if err != nil {
+		log.Print("Error while connect to ADB: ", err)
+		return
+	}
+	defer tcpConn.Close()
+
+	// Redirect WebSocket to ADB tcp socket
+	wsWrapper := &wsIoWrapper{
+		wsConn: wsConn,
+		pos: 0,
+		buf: nil,
+	}
+
+	go io.Copy(wsWrapper, tcpConn)
+	io.Copy(tcpConn, wsWrapper)
+}
+
+// Wrapper for implementing io.ReadWriteCloser of websocket.Conn
+type wsIoWrapper struct {
+	wsConn *websocket.Conn
+	pos int
+	buf []byte
+}
+
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
+
+func (w *wsIoWrapper) Read(buf []byte) (int, error) {
+	if w.buf == nil || w.pos == len(w.buf) {
+		_, buf, err := w.wsConn.ReadMessage()
```

Comment:
As we talked, the variable name `buf` seems duplicated.

Reply:
Changed buf name that is duplicated.